### PR TITLE
Add nonces to cart actions, ensure notices are visible for these actions

### LINKED ISSUE #19429: add_to_cart_nonce should be reinstated with a unique token for guests
Having a unique nonce on the add_to_cart form is what prevents XSRF attacks from being able to randomly add things to your cart. This used to be implemented, but was removed in this commit:
https://github.com/woocommerce/woocommerce/commit/5249935ce49055b91d980bba7048813e5e0ae606
The removal introduced a security hole. The reasoning was that the nonce was not unique for guest users so it was not safer anyway. While this is true, the correct fix would be to make a unique nonce, such as one based on the user's session ID used for their shopping cart. Removing it simply made an existing security hole bigger.

## Prerequisites

- [x] I have searched for similar issues in both open and closed tickets and cannot find a duplicate
- [x] The issue still exists against the latest `master` branch of WooCommerce on Github (this is **not** the same version as on WordPress.org!)
- [x] I have attempted to find the simplest possible steps to reproduce the issue
- [ ] I have included a failing test as a pull request (Optional)

## Steps to reproduce the issue

1. Start shopping. Add a couple of items to your basket.
2. Get caught by a XSRF attack, that adds a couple of items to your basket.
3. Go to checkout, pay for your items, without realising that there are more items than you were expecting. Lose money.

Eg. An attacker can simply submit a form from any domain like this:

<form id="foo" action="https://store.example.com/shop/some-item/"...>
<input type="hidden" name="attribute_size" value="L">
<input type="hidden" name="quantity" value="5">
<input type="hidden" name="add-to-cart" value="76" />
<input type="hidden" name="product_id" value="76" />
<input type="hidden" name="variation_id" class="variation_id" value="0" />
</form>
<script type="text/javascript">
document.getElementById('foo').submit();
</script>

## Expected/actual behavior

When I follow those steps, I see...

Additional items have been added to my cart, and I paid for them without realising.

I was expecting to see...

The XSRF attack should have been prevented.

## Isolating the problem

- [x] This bug happens with only WooCommerce plugin active
- [x] This bug happens with a default WordPress theme active, or [Storefront](https://woocommerce.com/storefront/)
- [x] I can reproduce this bug consistently using the steps above

## WordPress Environment

<details>
WOOCommerce Version 3.2.6
WOOCommerce Version 3.3.3
</details>

### PR DESCRIPTION
**Pull request by peterfabian**
Fixes #19429 

PR based on action items from [comment](https://github.com/woocommerce/woocommerce/issues/19429#issuecomment-378220238) on the issue.

## 1. Ensure add to cart events surface notices to users in all cases showing that an item has been added to cart.

- [x] `add_to_cart` and `remove_from_cart` notices added to /shop page when actions are performed using AJAX
- [x] single product page shows notice
- [x] https://www.yourwebsite.com/?add-to-cart=product_id shows notice
- [x] http://local.wordpress.test/shop/?add-to-cart=95 .. there is a notice
- [x] http://local.wordpress.test/cart/?add-to-cart=95 .. there is a notice
- [x] http://local.wordpress.test/checkout/?add-to-cart=95 .. there is a notice

## 2. Ensure if totals change or cart items are added during checkout, the checkout fails and a notice is displayed (I believe this happens already but will confirm)
- [x] added `woocommerce_cart_dirty` httponly cookie to flag when cart has been changed
- [x] added notice promptiong user to review order items when the cart is changed while in checkout
cart_updated

## 3. Include filter to add nonce checks to cart links, off by default.
- [x] added filter `woocommerce_add_to_cart_nonce_active`, off by default, affects adding to cart only
- [x] added nonce checks for AJAX `add_to_cart`, `remove_from_cart`
- [x] added nonce checks for non-AJAX `apply_coupon`, `remove_coupon` and `add-to-cart`
- [x] added nonces to links trigerring actions above

## 4. Include notice to confirm an add to cart event should the nonce be missing (simply forwards the link with nonce appended).
- [x] Notice added:
confirmation
 
- [x] Make sure it works for a selected variation of a variable product.

### CODE REVIEW

File: includes/class-wc-form-handler.php [hunk 1]
```diff
@@ -336,7 +336,13 @@ public static function checkout_action() {
 
 			wc_maybe_define_constant( 'WOOCOMMERCE_CHECKOUT', true );
 
-			WC()->checkout()->process_checkout();
+			// If cart has been modified since the user arrived to the checkout page and
+			// the page has not been updated, print notice for the user and don't process checkout.
+			if ( $_COOKIE['woocommerce_cart_dirty'] !== '0' && empty( $_POST['woocommerce_checkout_update_totals'] ) ) {
```
Comment (mikejolley) [inline includes/class-wc-form-handler.php]: I think `! empty` on the cart dirty cookie might be clearer and will avoid issues with int/string `0`.

File: includes/class-wc-form-handler.php [hunk 2]
```diff
@@ -336,7 +336,13 @@ public static function checkout_action() {
...
+				wc_add_notice( __( 'Cart has been updated. Please check your order before proceeding.', 'woocommerce' ), 'error' );
```
Comment (mikejolley) [inline includes/class-wc-form-handler.php]: `Your cart has been updated. Please check your cart contents before proceeding with your order.`

File: includes/class-wc-ajax.php [hunk 3]
```diff
@@ -396,6 +409,9 @@ public static function add_to_cart() {
 				wc_add_to_cart_message( array( $product_id => $quantity ), true );
 			}
 
+			$added_text = sprintf( __( '%s has been added to your cart.', 'woocommerce' ), $product->get_name() );
```
Comment (mikejolley) [inline includes/class-wc-ajax.php]: Why is this needed? It's immediately preceeded with `wc_add_to_cart_message`.

File: assets/js/frontend/add-to-cart.js [hunk 4]
```diff
@@ -5,6 +5,18 @@ jQuery( function( $ ) {
 		return false;
 	}
 
+	$.wc_clear_prepend_and_scroll_to_notices = function( notices, elementAfterNotices, scrollToElementSelector ) {
+		if ( notices ) {
```
Comment (mikejolley) [inline assets/js/frontend/add-to-cart.js]: Exit early - if there are no notices return here.

File: assets/js/frontend/add-to-cart.js [hunk 5]
```diff
@@ -5,6 +5,18 @@ jQuery( function( $ ) {
 		return false;
 	}
 
+	$.wc_clear_prepend_and_scroll_to_notices = function( notices, elementAfterNotices, scrollToElementSelector ) {
```
Comment (mikejolley) [inline assets/js/frontend/add-to-cart.js]: Does this function do too much?

Outputting the notices and scrolling to them could be a single method.

Clearing existing notices could be a separate method.

`wc_display_notices`/`wc_remove_notices`

File: includes/class-wc-ajax.php [hunk 6]
```diff
@@ -431,8 +457,48 @@ public static function remove_from_cart() {
 	 */
 	public static function checkout() {
 		wc_maybe_define_constant( 'WOOCOMMERCE_CHECKOUT', true );
-		WC()->checkout()->process_checkout();
-		wp_die( 0 );
+
+		if ( $_COOKIE['woocommerce_cart_dirty'] !== '0' ) {
+			wc_add_notice( __( 'Cart has been updated. Please check the order again.', 'woocommerce' ) );
+
+			// Get order review fragment
```
Comment (mikejolley) [inline includes/class-wc-ajax.php]: This is a lot of duplication from `update_order_review` - can they not share the same protected method to grab these fragments and return JSON?

File: includes/class-wc-ajax.php [hunk 7]
```diff
@@ -431,8 +457,48 @@ public static function remove_from_cart() {
 	 */
 	public static function checkout() {
 		wc_maybe_define_constant( 'WOOCOMMERCE_CHECKOUT', true );
-		WC()->checkout()->process_checkout();
-		wp_die( 0 );
+
+		if ( $_COOKIE['woocommerce_cart_dirty'] !== '0' ) {
```
Comment (mikejolley) [inline includes/class-wc-ajax.php]: Can you explain when this will be cleared in the event it's found here?

File: includes/class-wc-form-handler.php [hunk 8]
```diff
@@ -803,6 +809,31 @@ public static function add_to_cart_action( $url = false ) {
 			return;
 		}
 
+		$nonce_active = apply_filters( 'woocommerce_add_to_cart_nonce_active', false );
+		if ( $nonce_active ) {
+
+			if ( ! isset( $_REQUEST['woocommerce-cart-nonce'] ) && ! isset( $_REQUEST['_wpnonce'] ) ) {
+				//Include notice to confirm an add to cart event should the nonce be missing (simply forwards the link with nonce appended).
```
Comment (mikejolley) [inline includes/class-wc-form-handler.php]: Code standards? Is this comment in the right place?

File: includes/class-wc-form-handler.php [hunk 9]
```diff
@@ -803,6 +809,31 @@ public static function add_to_cart_action( $url = false ) {
...
+
+				// To get correct add_to_cart_url, we need variation object.
```
Comment (mikejolley) [inline includes/class-wc-form-handler.php]: This block should perhaps have it's own method - to display/get the confirmation notice and output it.

File: includes/class-wc-form-handler.php [hunk 10]
```diff
@@ -803,6 +809,31 @@ public static function add_to_cart_action( $url = false ) {
 			return;
 		}
 
+		$nonce_active = apply_filters( 'woocommerce_add_to_cart_nonce_active', false );
```
Comment (mikejolley) [inline includes/class-wc-form-handler.php]: Not keen on the naming of this hook. 

`woocommerce_add_to_cart_enable_nonce_check`? More descriptive IMO.

**Review by mikejolley — CHANGES_REQUESTED** [review]
Thanks @peterfabian. Added first round of feedback.

It might be better to split this PR into 2; one for add to cart nonce, and another for the other missing nonce checks. That might make this easier to review as the scope will be narrowed. The add to cart bits are more controversial :)

File: templates/single-product/add-to-cart/variation-add-to-cart-button.php [hunk 11]
```diff
@@ -24,6 +24,14 @@
 	) );
 
 	do_action( 'woocommerce_after_add_to_cart_quantity' );
+
+	/**
+	 * @since 3.4.0.
+	 */
+	$nonce_active = apply_filters( 'woocommerce_add_to_cart_nonce_active', false );
```
[the remark this answers is not in the document]
Comment (claudiulodro) [reply templates/single-product/add-to-cart/variation-add-to-cart-button.php]: +1 for doing this in a way that doesn't require a template update. The `before_` or `after_add_to_cart_button` hooks should work well.

File: includes/wc-cart-functions.php [hunk 12]
```diff
@@ -285,7 +285,7 @@ function wc_cart_totals_coupon_html( $coupon ) {
 	}
 
 	$discount_amount_html = apply_filters( 'woocommerce_coupon_discount_amount_html', $discount_amount_html, $coupon );
-	$coupon_html          = $discount_amount_html . ' <a href="' . esc_url( add_query_arg( 'remove_coupon', rawurlencode( $coupon->get_code() ), defined( 'WOOCOMMERCE_CHECKOUT' ) ? wc_get_checkout_url() : wc_get_cart_url() ) ) . '" class="woocommerce-remove-coupon" data-coupon="' . esc_attr( $coupon->get_code() ) . '">' . __( '[Remove]', 'woocommerce' ) . '</a>';
+	$coupon_html          = $discount_amount_html . ' <a href="' . esc_url( wp_nonce_url( add_query_arg( 'remove_coupon', rawurlencode( $coupon->get_code() ), defined( 'WOOCOMMERCE_CHECKOUT' ) ? wc_get_checkout_url() : wc_get_cart_url() ), 'woocommerce-cart' ) ) . '" class="woocommerce-remove-coupon" data-coupon="' . esc_attr( $coupon->get_code() ) . '">' . __( '[Remove]', 'woocommerce' ) . '</a>';
```
Comment (claudiulodro) [inline includes/wc-cart-functions.php]: There is a nonce already for [removing the coupon](https://github.com/woocommerce/woocommerce/blob/189ffab49103d9254a73edb0829094959a2a5b0f/includes/class-wc-frontend-scripts.php#L509). This should use that one instead of the Add to Cart nonce.

**Comment by mikejolley** [comment]
Punting this, unless time allows, since it needs work and we have API bits to prioritize.

**Comment by mikejolley** [comment]
@peterfabian Should we revisit this or start over, since it's stale and hasn't gained much attention?

**Comment by peterfabian** [comment]
I'd say let's close this one and give it a fresh look, maybe split it into smaller pieces first, what do you think, @mikejolley ?

**Comment by mikejolley** [comment]
Agreed. Probably too late for 3.6 now anyhow.
